p2p/transport/webrtc/listener.go

```diff
@@ -0,0 +1,310 @@
+				defer cancelFunc()
+				conn, err := l.accept(ctx, addr)
+				if err != nil {
+					log.Debugf("could not accept connection: %v", err)
+					return
+				}
+				l.connChan <- conn
+			}()
+		}
+	}
+}
+
+func (l *listener) Accept() (tpt.CapableConn, error) {
+	select {
+	case <-l.ctx.Done():
+		return nil, os.ErrClosed
+	case conn := <-l.connChan:
+		return conn, nil
+	}
+}
+
+func (l *listener) Close() error {
+	select {
+	case <-l.ctx.Done():
+		return nil
+	default:
+	}
+	l.cancel()
+	l.wg.Wait()
+	return nil
+}
+
+func (l *listener) Addr() net.Addr {
+	return l.mux.LocalAddr()
+}
+
+func (l *listener) Multiaddrs() []ma.Multiaddr {
+	return []ma.Multiaddr{l.localMultiaddr}
+}
+
+func (l *listener) accept(ctx context.Context, addr candidateAddr) (tpt.CapableConn, error) {
+	var (
+		scope network.ConnManagementScope
+		pc    *webrtc.PeerConnection
+	)
+
+	cleanup := func() {
+		if scope != nil {
+			scope.Done()
+		}
+		if pc != nil {
+			_ = pc.Close()
+		}
+	}
+
+	remoteMultiaddr, err := manet.FromNetAddr(addr.raddr)
+	if err != nil {
+		return nil, err
+	}
+
+	scope, err = l.transport.rcmgr.OpenConnection(network.DirInbound, false, remoteMultiaddr)
+	if err != nil {
+		defer cleanup()
+		return nil, err
+	}
+
+	settingEngine := webrtc.SettingEngine{}
+	settingEngine.SetAnsweringDTLSRole(webrtc.DTLSRoleServer)
+	settingEngine.SetICECredentials(addr.ufrag, addr.ufrag)
+	settingEngine.SetLite(true)
+	settingEngine.SetICEUDPMux(l.mux)
+	settingEngine.DisableCertificateFingerprintVerification(true)
+	settingEngine.DetachDataChannels()
+
+	api := webrtc.NewAPI(webrtc.WithSettingEngine(settingEngine))
+
+	pc, err = api.NewPeerConnection(l.config)
+	if err != nil {
+		defer cleanup()
+		return nil, err
+	}
+
+	// signaling channel wraps an error in a struct to make
+	// the error nullable.
+	signalChan := make(chan struct{ error })
+	var wrappedChannel *dataChannel
+	var handshakeOnce sync.Once
+	// this enforces that the correct data channel label is used
+	// for the handshake
+	handshakeChannel, err := pc.CreateDataChannel("handshake", &webrtc.DataChannelInit{
+		Negotiated: func(v bool) *bool { return &v }(true),
+		ID:         func(v uint16) *uint16 { return &v }(0),
+	})
+	if err != nil {
+		defer cleanup()
+		return nil, err
+	}
+
+	// The raw datachannel is wrapped in the libp2p abstraction
+	// as early as possible to allow any messages sent by the remote
+	// to be buffered. This is done since the dialer leads the listener
+	// in the handshake process, and a faster dialer could have set up
+	// their connection and started sending Noise handshake messages before
+	// the listener has set up the onmessage callback. In this use case,
+	// since the data channels are negotiated out-of-band, they will be
+	// instantly in `readyState=open` once the SCTP connection is set up.
+	// Therefore, we wrap the datachannel before performing the
+	// offer-answer exchange, so any messages sent from the remote get
+	// buffered.
```

Comment:
Sorry, this comment seems to be from when we were using the onmessage callback. What happens is that if the datachannel was not wrapped by our struct, the onmessage callback for messages would not be called and would cause messages to be dropped. This effectively created a race condition between the remote sending data and the local datachannel being wrapped and having the onmessage callback and buffer set up.